apps/api/src/persistence/repositories/ArangoCategoryRepository.ts

```diff
@@ -0,0 +1,120 @@
+import categoryValidator from '../../domain/domainModelValidators/categoryValidator';
+import buildInstanceFactory from '../../domain/factories/utilities/buildInstanceFactory';
+import { Category } from '../../domain/models/categories/entities/category.entity';
+import { ICategoryRepository } from '../../domain/repositories/interfaces/ICategoryRepository';
+import { EntityId } from '../../domain/types/ResourceId';
+import { InternalError } from '../../lib/errors/InternalError';
+import { Maybe } from '../../lib/types/maybe';
+import { isNotFound, NotFound } from '../../lib/types/not-found';
+import { DTO } from '../../types/DTO';
+import { ArangoDatabase } from '../database/arango-database';
+import { DatabaseProvider } from '../database/database.provider';
+import {
+    categoryCollectionID,
+    categoryEdgeCollectionID,
+} from '../database/types/ArangoCollectionId';
+import { HasArangoDocumentDirectionAttributes } from '../database/types/HasArangoDocumentDirectionAttributes';
+import convertArangoDocumentHandleToCompositeIdentifier from '../database/utilities/convertArangoDocumentHandleToCompositeIdentifier';
+import mapDatabaseDTOToEntityDTO from '../database/utilities/mapDatabaseDTOToEntityDTO';
+import { DatabaseDocument } from '../database/utilities/mapEntityDTOToDatabaseDTO';
+
+type CategoryDocument = Omit<DatabaseDocument<DTO<Category>>, 'childrenIDs'>;
+
+type LookupTableForChildrenIDs = Map<EntityId, EntityId[]>;
+
+export default class ArangoCategoryRepository implements ICategoryRepository {
+    #instanceFactory = buildInstanceFactory(categoryValidator, Category);
+
+    #arangoDB: ArangoDatabase;
+
+    constructor(arangoDatabaseProvider: DatabaseProvider) {
+        this.#arangoDB = arangoDatabaseProvider.getDBInstance();
+    }
+
+    async fetchTree(): Promise<(Category | InternalError)[]> {
+        const categories = await this.#arangoDB.fetchMany<CategoryDocument>(categoryCollectionID);
+
+        const category_edges = await this.#arangoDB.fetchMany<HasArangoDocumentDirectionAttributes>(
+            categoryEdgeCollectionID
+        );
+
+        const categoryDTOs = this.#buildCategoryDTOs(categories, category_edges);
+
+        return categoryDTOs.map((dto) => this.#instanceFactory(dto));
+    }
+
+    async fetchById(id: EntityId): Promise<Maybe<Category | InternalError>> {
+        const categoryDocument = await this.#arangoDB.fetchById<CategoryDocument>(
+            id,
+            categoryCollectionID
+        );
+
+        if (isNotFound(categoryDocument)) return NotFound;
+
+        const edgesWhereThisCategoryIsTheParent = await this.#arangoDB
+            .fetchMany<HasArangoDocumentDirectionAttributes>(categoryEdgeCollectionID)
```

Comment:
It might be nice to optionally add some filters as part of the `fetchMany` call. Haven't thought about how this would relate to our `Specification translation layer`.